planning/planning_validator/include/planning_validator/utils.hpp

```diff
@@ -0,0 +1,57 @@
+// Copyright 2021 Tier IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef PLANNING_VALIDATOR__UTILS_HPP_
+#define PLANNING_VALIDATOR__UTILS_HPP_
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <autoware_auto_planning_msgs/msg/trajectory.hpp>
+
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace planning_validator
+{
+using autoware_auto_planning_msgs::msg::Trajectory;
+using autoware_auto_planning_msgs::msg::TrajectoryPoint;
+
+std::pair<double, size_t> getMaxValAndIdx(const std::vector<double> & v);
+
+std::pair<double, size_t> getMinValAndIdx(const std::vector<double> & v);
+
+std::pair<double, size_t> getAbsMaxValAndIdx(const std::vector<double> & v);
+
+Trajectory resampleTrajectory(const Trajectory & trajectory, const double min_interval);
+
+std::vector<double> calcCurvature(const Trajectory & trajectory);
+
+std::vector<double> calcIntervalDistance(const Trajectory & trajectory);
+
+std::vector<double> calcLateralAcceleration(const Trajectory & trajectory);
+
+std::vector<double> getLongitudinalAccArray(const Trajectory & trajectory);
+
+std::vector<double> calcRelativeAngles(const Trajectory & trajectory);
+
+std::vector<double> calcSteeringAngles(const Trajectory & trajectory, const double wheelbase);
+
+std::vector<double> calcSteeringRates(const Trajectory & trajectory, const double wheelbase);
```

Comment:
It would be better to pass it as a smart pointer or argument since memory copying may occur.

Reply:
Thank you for the comment. However, using smart pointer or reference argument just for this purpose is not preferred by me due to the readability. So many discussions for NRVO and no clear answer yet. I decided to leave it to optimization, and believe that is better in almost all cases recently.

Reply:
I understand how you feel.
It would be nice if NVRO was guaranteed, but currently it is no.
Because of those limitations, many codes employ argument passing by reference.

http://opencv.jp/opencv-2.1/cpp/geometric_image_transformations.html
http://docs.ros.org/en/indigo/api/tf/html/c++/classtf_1_1Transformer.html

Reply:
@yukkysaito @TakaHoribe 
(Sorry for long message, but this is translated by DeepL. )

In my opinion, readability should be prioritized at first because the overhead gives too small influence to Autoware's functionality.

I tried measuring elapsed time on both "returning an object" and "returning a pointer" briefly .  When a function returns a Trajectory object which has 1000 trajectory points, about 10~20 "micro"seconds cost.  
On the other hand, returning pointer costs  about 50 ~ 200 "nano"seconds.   
I found that "returning the object " sometimes costs 100 times as much time as "returning the pointer".  
As I'm not sure how large data `Trajectory` object has, it might not be correct. 

There is exactly a large gap between returning the object and returning the pointer.  
I think that this impact is too small to influence on Autwoare's functionality if `resampleTrajectory()` is called at 10 Hz. If it is called frequently like at 1000 Hz, we may need to take care of the overhead of returning the object.

My opinion came from my bias, but I think other opinions should be taken care of.

Reply:
@takam5f2 @yukkysaito Thank you for your comment. I've tested how the computational time is improved. Changing all implementation to reference argument improves 0.006 [ms] on average for the whole process in 500 times operation with 166 points trajectory (trajectory usually has about 100~200 points). Even with 1500 points trajectory, the improvement was 0.012 [ms]. Readability should not be sacrificed for this small improvement.

>  if NRVO was guaranteed, but currently it is no.

I know the background. Even if it is not guaranteed, it is optimized anyway. It should be ok.

> Also, allowing this would allow the point cloud to be the return value.

I totally doubt that. It should depend on the implementation. We definitely shouldn't make a copy just to change a few fields in a large object.

Well, we shouldn't take time in this discussion in this minor thread. Actually, there are already many implementations using "return large object" in the autoware. If you really care about this point, please make a discussion thread and put the decision into the coding guideline.

Here I'll take the new implementation to skip the discussion.

Reply:
@takam5f2 @yukkysaito Sorry I misunderstood. ["Using reference argument](https://github.com/autowarefoundation/autoware.universe/pull/1947/commits/e81c91bafc0e61eaa9b6fa63feabba96205470ff)" is 0.006 [ms] SLOWER than "[returning object](https://github.com/autowarefoundation/autoware.universe/pull/1947/commits/f36c7820ba47ccd3fbcd614e0aca0c414750b9cf)".

Reply:
Thank you. Memory access is more of a concern than execution time, as memory access is currently tight on TIER IV computers.

>I know the background. Even if it is not guaranteed, it is optimized anyway. It should be ok.

Sorry. Please tell me what meaning? :pray: 
The optimization is running in the current environment, but if we change the code slightly or environment, it may not be optimized.

Reply:
> If you really care about this point, please make a discussion thread and put the decision into the coding guideline.

Yes. Let's talk separately.

Reply:
Commented here: https://github.com/orgs/autowarefoundation/discussions/3209#discussioncomment-4723928